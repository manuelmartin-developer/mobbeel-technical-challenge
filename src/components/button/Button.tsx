import styles from "./Button.module.scss";

interface ButtonProps {
  /**
   * The text to display in the button
   * @type {string}
   * @memberof ButtonProps
   * @required
   * @example
   * text="Next"
   *
   * */
  text: string | React.ReactNode;

  /**
   * The function to call when the button is clicked
   * @type {() => void}
   * @memberof ButtonProps
   * @required
   * @example
   * onClick={() => console.log("Button clicked")}
   *
   * */
  onClick: () => void;

  /**
   * Whether the button is disabled
   * @type {boolean}
   * @memberof ButtonProps
   * @example
   * disabled={true}
   *
   * */
  disabled?: boolean;

  /**
   * Whether the button is an icon button
   * @type {boolean}
   * @memberof ButtonProps
   * @example
   * iconButton={true}
   *
   * */
  iconButton?: boolean;

  /**
   * The width of the button
   * @type {string}
   * @memberof ButtonProps
   * @example
   * width="100px"
   *
   * */
  width?: string;

  /**
   * Whether the button is selected
   * @type {boolean}
   * @memberof ButtonProps
   * @example
   * selected={true}
   *
   * */
  height?: string;

  /**
   * Whether the button is selected
   * @type {boolean}
   * @memberof ButtonProps
   * @example
   * selected={true}
   *
   * */
  selected?: boolean;

  /**
   * The test id of the button
   * @type {string}
   * @memberof ButtonProps
   * @example
   * testId="button"
   * @default button
   * */
  testId?: string;
}

const Button: React.FC<ButtonProps> = ({
  text,
  onClick,
  disabled,
  iconButton,
  width,
  height,
  selected,
  testId = "button",
}) => {
  return (
    <button
      data-testid={testId}
      className={`${styles.button} ${disabled ? styles.button_disabled : ""} ${iconButton ? styles.button_icon : ""} ${selected ? styles.button_selected : ""}
      `}
      role="button"
      onClick={onClick}
      aria-disabled={disabled}
      style={{ width, height }}
    >
      <span>{text}</span>
    </button>
  );
};

export default Button;
