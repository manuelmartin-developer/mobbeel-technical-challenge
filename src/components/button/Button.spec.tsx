import "@testing-library/jest-dom";
import { describe, expect, it } from "vitest";
import { render, screen } from "@testing-library/react";

import Button from "./Button";

describe("Button", () => {
  it("renders Button component", () => {
    render(<Button text="Button" onClick={() => null} />);
    expect(screen.getByTestId("button")).toBeInTheDocument();
  });

  it("renders text", () => {
    render(<Button text="Button" onClick={() => null} />);
    expect(screen.getByText("Button")).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const { asFragment } = render(
      <Button text="Button" onClick={() => null} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot with disabled", () => {
    const { asFragment } = render(
      <Button text="Button" onClick={() => null} disabled />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot with iconButton", () => {
    const { asFragment } = render(
      <Button text="Button" onClick={() => null} iconButton />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot with width", () => {
    const { asFragment } = render(
      <Button text="Button" onClick={() => null} width="100px" />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot with height", () => {
    const { asFragment } = render(
      <Button text="Button" onClick={() => null} height="100px" />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot with selected", () => {
    const { asFragment } = render(
      <Button text="Button" onClick={() => null} selected />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
