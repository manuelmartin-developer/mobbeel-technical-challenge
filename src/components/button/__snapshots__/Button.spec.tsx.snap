// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="_button_a9a5cd   
      "
    data-testid="button"
    role="button"
  >
    <span>
      Button
    </span>
  </button>
</DocumentFragment>
`;

exports[`Button > matches snapshot with disabled 1`] = `
<DocumentFragment>
  <button
    aria-disabled="true"
    class="_button_a9a5cd _button_disabled_a9a5cd  
      "
    data-testid="button"
    role="button"
  >
    <span>
      Button
    </span>
  </button>
</DocumentFragment>
`;

exports[`Button > matches snapshot with height 1`] = `
<DocumentFragment>
  <button
    class="_button_a9a5cd   
      "
    data-testid="button"
    role="button"
    style="height: 100px;"
  >
    <span>
      Button
    </span>
  </button>
</DocumentFragment>
`;

exports[`Button > matches snapshot with iconButton 1`] = `
<DocumentFragment>
  <button
    class="_button_a9a5cd  _button_icon_a9a5cd 
      "
    data-testid="button"
    role="button"
  >
    <span>
      Button
    </span>
  </button>
</DocumentFragment>
`;

exports[`Button > matches snapshot with selected 1`] = `
<DocumentFragment>
  <button
    class="_button_a9a5cd   _button_selected_a9a5cd
      "
    data-testid="button"
    role="button"
  >
    <span>
      Button
    </span>
  </button>
</DocumentFragment>
`;

exports[`Button > matches snapshot with width 1`] = `
<DocumentFragment>
  <button
    class="_button_a9a5cd   
      "
    data-testid="button"
    role="button"
    style="width: 100px;"
  >
    <span>
      Button
    </span>
  </button>
</DocumentFragment>
`;
